### PR TITLE
[Bilibili] fix BilibiliSpaceBaseIE with playlist-items

### PR DESCRIPTION
**Pull request by lockmatrix**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

- fix BilibiliSpaceBaseIE with playlist-items
- refine BilibiliSpaceVideoIE message when being blocked
  - the whole message refine needs https://github.com/yt-dlp/yt-dlp/pull/4945

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Comment by lockmatrix** [comment]
> All of these features already exist in core. See --sleep-requests and --playlist-items

`--sleep-requests` is this function.

while `--playlist-items` seems different, it can skip download progress, but `start_page,end_page` is also for skip playlist page fetching

**Comment by pukkandan** [comment]
Assuming the extractor is written correctly, `--playlist-items` causes only the necessary playlist pages to be downloaded. If this is not working for this case, you may need to change the `entries` returned to an `InAdvancePagedList`

**Comment by pukkandan** [comment]
> If this is not working for this case, you may need to change the `entries` returned to an `InAdvancePagedList`

Skimming through the code, this does seem to be already being used. Show log of playlist-items not working

**Comment by lockmatrix** [comment]
> Assuming the extractor is written correctly, --playlist-items causes only the necessary playlist pages to be downloaded. If this is not working for this case, you may need to change the entries returned to an InAdvancePagedList

get, I will try it.

**Comment by lockmatrix** [comment]
> Skimming through the code, this does seem to be already being used. Show log of playlist-items not working

```
yt-dlp.exe -vU --cookies-from-browser chrome  --sleep-requests 2 --playlist-items 100:110 https://space.bilibili.com/6444277/video
[debug] Command-line config: ['-vU', '--cookies-from-browser', 'chrome', '--sleep-requests', '2', '--playlist-items', '100:110', 'https://space.bilibili.com/6444277/video']
[debug] Encodings: locale cp936, fs utf-8, pref cp936, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.09.17.015137 [e02485b] (win32_exe)
[debug] Python 3.8.10 (CPython 64bit) - Windows-10-10.0.19044-SP0
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg N-108089-gc92edd969a-20220910 (setts), ffprobe N-108089-gc92edd969a-20220910, phantomjs 2.1.1
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.09.14, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[Cookies] Extracting cookies from chrome
[debug] Extracting cookies from: "C:\Users\haman\AppData\Local\Google\Chrome\User Data\Default\Network\Cookies"
[debug] Found local state file at "C:\Users\haman\AppData\Local\Google\Chrome\User Data\Local State"
[Cookies] Extracted 3280 cookies from chrome
[debug] cookie version breakdown: {'v10': 3280, 'other': 0, 'unencrypted': 0}
[debug] Proxy map: {}
[debug] Loaded 1678 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.09.01, Current version: 2022.09.17.015137
yt-dlp is up to date (2022.09.17.015137)
[debug] [BilibiliSpaceVideo] Extracting URL: https://space.bilibili.com/6444277/video
[BilibiliSpaceVideo] 6444277: Downloading page 1
[download] Downloading playlist: 6444277
[BilibiliSpaceVideo] Sleeping 2.0 seconds ...
[BilibiliSpaceVideo] 6444277: Downloading page 3
[BilibiliSpaceVideo] Playlist 6444277: Downloading 11 videos
[download] Downloading video 1 of 11
[debug] [BiliBili] Extracting URL: https://www.bilibili.com/video/BV1B5411Q7Fz
[BiliBili] Sleeping 2.0 seconds ...
[BiliBili] 1B5411Q7Fz: Downloading webpage
[BiliBili] Sleeping 2.0 seconds ...
[BiliBili] 1B5411Q7Fz: Extracting videos in anthology
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[BiliBili] Sleeping 2.0 seconds ...
```

It looks weird:
* download page 1
* download page 3
* start download video ....

**Comment by pukkandan** [comment]
The first page is needed for getting metadata https://github.com/yt-dlp/yt-dlp/blob/0ca0f88121db5a1e9c223077af1b78c62d5ead6d/yt_dlp/extractor/bilibili.py#L511-L512

**Comment by lockmatrix** [comment]
> The first page is needed for getting metadata

I know it, 

I found the page_size is 30, not 10, so it should download page 4, not 3.
Is InAdvancePagedList assume idx start from 0?

**Comment by pukkandan** [comment]
> Is InAdvancePagedList assume idx start from 0?

Yes

**Comment by lockmatrix** [comment]
> Yes

en, I need to fix it

**Comment by lockmatrix** [comment]
@pukkandan I finished.
Maybe I should split "BilibiliSpaceBaseIE page_start_idx fix" to other PR?

File: yt_dlp/extractor/bilibili.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 
 from .common import InfoExtractor, SearchInfoExtractor
 from ..compat import (
+    compat_HTTPError,
```
Comment (pukkandan) [inline yt_dlp/extractor/bilibili.py:10]: urllib.error.HTTPError

File: yt_dlp/extractor/bilibili.py [hunk 2]
```diff
@@ -535,10 +536,19 @@ def _real_extract(self, url):
...
+                                               query={'mid': playlist_id, 'pn': page_idx + 1, 'jsonp': 'jsonp'})
+            except ExtractorError as e:
+                if isinstance(e.cause, compat_HTTPError) and e.cause.code == 412:
+                    raise ExtractorError('Request is blocked by server, please add cookies, wait and try later.', expected=True)
+                raise
+            if response['code'] == -401:
+                self.to_screen(f'response: {response}')
+                raise ExtractorError('Request is blocked by server, please add cookies, wait and try later.', expected=True)
+            data = response['data']
+            return data
```
Comment (pukkandan) [inline yt_dlp/extractor/bilibili.py:551]:
```suggestion
            return response['data']
```

**Review by pukkandan — CHANGES_REQUESTED** [review]
> Maybe I should split "BilibiliSpaceBaseIE page_start_idx fix" to other PR?

No, this PR is fine as-is

File: yt_dlp/extractor/bilibili.py [hunk 3]
```diff
@@ -535,10 +536,19 @@ def _real_extract(self, url):
...
+            try:
+                response = self._download_json('https://api.bilibili.com/x/space/arc/search',
+                                               playlist_id, note=f'Downloading page {page_idx}',
+                                               query={'mid': playlist_id, 'pn': page_idx + 1, 'jsonp': 'jsonp'})
+            except ExtractorError as e:
+                if isinstance(e.cause, compat_HTTPError) and e.cause.code == 412:
+                    raise ExtractorError('Request is blocked by server, please add cookies, wait and try later.', expected=True)
+                raise
+            if response['code'] == -401:
+                self.to_screen(f'response: {response}')
```
[the remark this answers is not in the document]
Comment (lockmatrix) [reply yt_dlp/extractor/bilibili.py:548]: get

**Comment by lockmatrix** [comment]
@pukkandan 
I finished.

**Comment by pukkandan** [comment]
There is no need to ping me every time

File: yt_dlp/extractor/bilibili.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pukkandan) [reply yt_dlp/extractor/bilibili.py]: You didnt add anything to the message?

File: yt_dlp/extractor/bilibili.py [hunk 4]
```diff
@@ -535,10 +536,17 @@ def _real_extract(self, url):
...
+            try:
+                response = self._download_json('https://api.bilibili.com/x/space/arc/search',
+                                               playlist_id, note=f'Downloading page {page_idx}',
+                                               query={'mid': playlist_id, 'pn': page_idx + 1, 'jsonp': 'jsonp'})
+            except ExtractorError as e:
+                if isinstance(e.cause, urllib.error.HTTPError) and e.cause.code == 412:
+                    raise ExtractorError('Request is blocked by server, please add cookies, wait and try later.', expected=True)
+                raise
+            if response['code'] == -401:
+                raise ExtractorError('Request is blocked by server, please add cookies, wait and try later.', expected=True)
```
Comment (pukkandan) [inline yt_dlp/extractor/bilibili.py:548]:
```suggestion
                    raise ExtractorError(
                        'Request is blocked by server (412), please add cookies, wait and try later.', expected=True)
                raise
            if response['code'] == -401:
                raise ExtractorError(
                    'Request is blocked by server (401), please add cookies, wait and try later.', expected=True)
```

**Pull request merged**